### ISSUE TITLE
HID example does not work on Radxa Rockpi 4B+ (dwc3)

### ISSUE THREAD

**Issue by herbertp** [vpelletier/python-functionfs#28]
When trying to run the HID example on a recent version of Armbian, everything seems to be fine from the Gadget/python side, but the Host side reports the following errors:

```
[11529.096555] usb 2-1: new SuperSpeed USB device number 18 using xhci_hcd
[11529.117158] usb 2-1: no configurations
[11529.117182] usb 2-1: can't read configurations, error -22
[11529.320528] usb 2-1: new SuperSpeed USB device number 19 using xhci_hcd
[11529.341146] usb 2-1: no configurations
[11529.341167] usb 2-1: can't read configurations, error -22
[11529.349908] usb usb2-port1: attempt power cycle
[11530.276543] usb 2-1: new SuperSpeed USB device number 20 using xhci_hcd
[11530.297214] usb 2-1: no configurations
[11530.297236] usb 2-1: can't read configurations, error -22
[11530.500545] usb 2-1: new SuperSpeed USB device number 21 using xhci_hcd
[11530.521191] usb 2-1: no configurations
[11530.521212] usb 2-1: can't read configurations, error -22
[11530.529882] usb usb2-port1: unable to enumerate USB device
```

Note that the kernel_ncm example works just fine and gives:
```
[537255.940114] usb 2-1: new SuperSpeed USB device number 22 using xhci_hcd
[537255.960961] usb 2-1: New USB device found, idVendor=1d6b, idProduct=0104, bcdDevice= 6.01
[537255.960990] usb 2-1: New USB device strings: Mfr=0, Product=0, SerialNumber=0
[537256.028546] usbcore: registered new interface driver cdc_ether
[537256.061476] cdc_ncm 2-1:1.0: MAC-Address: 0a:fb:ea:89:33:bf
[537256.062605] cdc_ncm 2-1:1.0 usb0: register 'cdc_ncm' at usb-0000:01:00.0-1, CDC NCM (NO ZLP), 0a:fb:ea:89:33:bf
[537256.063443] usbcore: registered new interface driver cdc_ncm
[537256.071446] usbcore: registered new interface driver cdc_wdm
[537256.077169] usbcore: registered new interface driver cdc_mbim
```

System:
Armbian 23.02.0-trunk Lunar with bleeding edge Linux 6.1.24-rockchip64.
Rockpi 4Bplus with USB port configured to peripherial.
usb_f_fs and libcomposite modules loaded.
configfs mounted.

**Comment by herbertp** [comment]
Based on the suggestions in issue #29 I tried a few things and here are the results:

- Tested with mainline 6.2.5 (will test with newer version soon) and got the same issue.
- Tested with an USB 2.0 (host), there is works regardless of the kernel.
- Disconnected and reconnected the "running" gadget to an USB 3.x host and it worked as well.

Note: on USB 2.0 the gadget reports "onEnable called" while on 3.x no message is issued.

My Setup is an RPi4 acting as Host (2.0 and 3.x) and the RockPi 4B+ as (3.x) peripheral.
My conclusion so far is that the example does not work with USB 3.x host and gadget while it seems to work fine when limited to USB 2.0.

Would be great to get some confirmation that the example works on USB 3.x capable systems.

Thanks,
Herbert

**Comment by vpelletier** [comment]
I think I have a fix: 36bd59f15bd90bf2fddbdec369d2aa6e376ed082 .

Could you please try this change on real hardware ?

**Comment by herbertp** [comment]
Unfortunately no luck on real hardware ...

RPi4 reports:
```
[4186901.631937] usb 2-1: new SuperSpeed USB device number 35 using xhci_hcd
[4186901.652502] usb 2-1: no configurations
[4186901.652529] usb 2-1: can't read configurations, error -22
[4186901.855913] usb 2-1: new SuperSpeed USB device number 36 using xhci_hcd
[4186901.876449] usb 2-1: no configurations
[4186901.876472] usb 2-1: can't read configurations, error -22
[4186901.885154] usb usb2-port1: attempt power cycle
[4186902.811914] usb 2-1: new SuperSpeed USB device number 37 using xhci_hcd
[4186902.832528] usb 2-1: no configurations
[4186902.832550] usb 2-1: can't read configurations, error -22
[4186903.035927] usb 2-1: new SuperSpeed USB device number 38 using xhci_hcd
[4186903.056526] usb 2-1: no configurations
[4186903.056547] usb 2-1: can't read configurations, error -22
[4186903.065184] usb usb2-port1: unable to enumerate USB device
```
... and the RockPi (Peripheral) doesn't get beyond ...
```
Gadget ready, waiting for function to exit.
```


Thanks,
Herbert

**Comment by herbertp** [comment]
Small update here, I've built a 6.3.9 kernel (for the RockPi) which shows exactly the same behavior:
```
[4293194.881651] usb usb2-port1: attempt power cycle
[4293195.393110] usb 2-1: new SuperSpeed USB device number 49 using xhci_hcd
[4293195.413689] usb 2-1: no configurations
[4293195.413715] usb 2-1: can't read configurations, error -22
[4293195.617154] usb 2-1: new SuperSpeed USB device number 50 using xhci_hcd
[4293195.637702] usb 2-1: no configurations
[4293195.637727] usb 2-1: can't read configurations, error -22
[4293195.646453] usb usb2-port1: unable to enumerate USB device
```
... and again, it works fine if connected to an USB 2.0 only port.

Hope this helps,
Herbert

**Comment by herbertp** [comment]
Actually I take it all back, turns out that I updated the git repo, but forgot to install python-functionfs ...
With the updated version installed, it now actually works on real hardware as well!

Thanks again for the quick fix!

**Comment by vpelletier** [comment]
> With the updated version installed, it now actually works on real hardware as well!

Great, I was wondering if there was another layer to this issue, with the UDC, and was about to suggest that you repeat the test I did with dummy_udc on your device (which, it turns out, would have caught the missed update).

I added a docstring-only fix on top of the code change and released the result as 0.9.2 .

Thank you very much for your tests !
